for cmake3 make tbb an imported target

pwm1234 opened the pull request:
addresses an error when exporting a target that depends on tbb

I described this error with more detail in #5. I am by no means a cmake expert, but I needed this change for my work, and I wanted to contribute back since I have found findTBB.cmake helpful.

pwm1234-sri committed:
for cmake3 make tbb an imported target

addresses an error when exporting a target that depends on tbb

justusc reviewed (changes requested):
Thanks for the pull request.

The initial changes look OK, but similar changes will also need to be applied to the other branches below.

Did you by any chance run the test project?

pwm1234 commented:
No, I did not run the test project. Let me do that and let you know how things go. I just ran into some weirdness that I need to investigate, so put this pull request on hold for now. (Is there a formal way to do that?) I will send you a note when I have tried the test and think it is ready.
Phil

justusc commented:
Exporting external targets into a config file is a non-trivial process. The documentation on the procedure is very sparse. The trick is that you have to recreate the external targets in your exported config file. The recommended procedure is: (1) export the find module into the same directory as your config file, and (2) write your config file to use the find module to recreate the external target (i.e. the TBB target). Another approach is to use the TBB _variables_ to set the appropriate compile options for your exported libraries. Unfortunately the second option precludes using the TBB target in your project as it will result in errors.

The test project is just the project I used to ensure that FindTBB is working as expected and can be used to build an executable. To run the test project run `cmake .` in the test directory.

Try modifying the test project to be a minimal example of what you are trying to do in your project.

I will look into this issue in more detail later this week, if you don't beat me to it.

pwm1234-sri committed:
add install export that demonstrates problem

Because FindTbb makes tbb a real target, using target_link_libraries
with tbb generates the following error when using install export

CMake Error: install(EXPORT "depends" ...) includes target "test_tbb_c" which requires target "tbb" that is not in the export set.

pwm1234-sri committed:
fix problem with making tbb an imported target

use IMPORTED_LOCATION_<CONFIG> rather than INTERFACE_LINK_LIBRARIES

pwm1234 commented:
Well, I gave it my best effort. The problem with my first attempt was I used INTERFACE_LINK_LIBRARIES for an imported target, which is not correct. I just pushed a commit to my pull request where I use `IMPORTED_LOCATION_<CONFIG>`, which I think is the correct way to configure an imported target. I also added to the test project install export rules that demonstrate the original error that made me go down this path in the first place. 

I tried to put in changes for the other two branches (where (1) only TBB_LIBRARIES_RELEASE is defined, or (2) neither TBB_LIBRARIES_RELEASE nor TBB_LIBRARIES_DEBUG are defined), but that logic seems a little strange and testing it with msvc14, which always defines both, seems to take more time than I have right now.

I hope this PR is helpful.

justusc commented:
I think you are going in the right direction, but exporting targets in CMake is a tricky business. In my own projects, I just used the TBB variables directly to create executables and libraries in config files. I will try to get it working as soon as I have some free time. Hopefully this weekend.

pwm1234 commented:
I think the focus of this issue and the pull request is to change the tbb target from being an INTERFACE target to being an IMPORTED target. (This is what FindBoost.cmake does in cmake 3.6). 

Yes, I understand that exporting targets in CMake is tricky; I admit that I understand little about it, so I appreciate you taking time to look at this. Also, by no means do I think my pull request has a good example of exporting a target. It is just a demonstration of the error I encountered when exporting a target that used the tbb target.

justusc commented:
The changes you made in your PR were mostly correct. I made a minor change to the FindTBB.cmake file. Otherwise, everything looked good.

I also expanded the test to verify that the FindTBB script can be used by projects that export executables and shared libraries that depend on TBB.

pwm1234 commented:
Thank you! I appreciate you making FindTBB available and your time incorporating this request.
